src/streaming.ts

```diff
@@ -0,0 +1,71 @@
+import type { CompletionResponse } from './schemas/completion';
+
+/**
+ * Convenience type for brevity in declarations.
+ */
+type AugmentedCompletionResponse = {
+  completion: string;
+  response: CompletionResponse;
+};
+
+/**
+ * A parser for the streaming responses from the OpenAI API.
+ *
+ * Conveniently shaped like an argument for WritableStream constructor.
+ */
+class OpenAIStreamParser {
+  onchunk?: (chunk: AugmentedCompletionResponse) => void;
+  onend?: () => void;
+
+  /**
+   * Takes the ReadableStream chunks, produced by `fetch` and turns them into
+   * `CompletionResponse` objects.
+   * @param chunk The chunk of data from the stream.
+   */
+  write(chunk: Uint8Array): void {
+    const decoder = new TextDecoder();
+    const s = decoder.decode(chunk);
+    s.split('\n')
+      .map((line) => line.trim())
+      .filter((line) => line.length > 0)
+      .forEach((line) => {
+        const pos = line.indexOf(':');
+        const name = line.substring(0, pos);
+        if (name !== 'data') return;
+        const content = line.substring(pos + 1).trim();
+        if (content.length == 0) return;
+        if (content === '[DONE]') {
+          this.onend?.();
+          return;
+        }
+        const parsed = JSON.parse(content);
+        this.onchunk?.({
+          completion: parsed.choices[0].text || '',
```

Comment:
I think you should just pass the parsed content directly instead of adding additional metadata

this.onchunk?.(parsed)

Reply:
I wanted to make sure this stays in spirit with the rest of the API, so I kept the shape of the object the same as the response in non-streaming completion.

Reply:
Yeah, I think it's best to keep it consistent with the rest of the API.